core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -289,64 +309,105 @@ Path versionHintFile() {
     return metadataPath(Util.VERSION_HINT_FILENAME);
   }
 
-  private void writeVersionHint(int versionToWrite) {
+  @VisibleForTesting
+  void writeVersionHint(FileSystem fs, Integer versionToWrite) throws Exception {
     Path versionHintFile = versionHintFile();
-    FileSystem fs = getFileSystem(versionHintFile, conf);
-
+    Path tempVersionHintFile = metadataPath(UUID.randomUUID() + "-version-hint.temp");
     try {
-      Path tempVersionHintFile = metadataPath(UUID.randomUUID().toString() + "-version-hint.temp");
       writeVersionToPath(fs, tempVersionHintFile, versionToWrite);
-      fs.delete(versionHintFile, false /* recursive delete */);
       fs.rename(tempVersionHintFile, versionHintFile);
-    } catch (IOException e) {
-      LOG.warn("Failed to update version hint", e);
+    } catch (Exception e) {
+      // Cleaning up temporary files.
+      if (fs.exists(tempVersionHintFile)) {
+        io().deleteFile(tempVersionHintFile.toString());
+      }
+      throw e;
+    }
+  }
+
+  @VisibleForTesting
+  void deleteOldVersionHint(FileSystem fs, Path versionHintFile, Integer nextVersion)
+      throws IOException {
+    if (fs.exists(versionHintFile)) {
+      if (versionHintIsCorrupted(fs, versionHintFile)) {
+        throw new RuntimeException("VersionHint is corrupted! We will reject this commit.");
+      }
+      if (!fs.delete(versionHintFile, false /* recursive delete*/)) {
+        String msg =
+            String.format(
+                "Can not drop old versionHint. commitVersion = %s.Do you have multiple tasks working on this table at the same time?",
+                nextVersion);
+        throw new RuntimeException(msg);
+      }
     }
   }
 
-  private void writeVersionToPath(FileSystem fs, Path path, int versionToWrite) throws IOException {
+  @VisibleForTesting
+  boolean versionHintIsCorrupted(FileSystem fs, Path versionHintFile) throws IOException {
```

Comment:
This is very misleading. The version hint is not corrupted. The version hint is just not up to date.